File: trabalho3/Sistema/views/cadastro/passo3.py
"""Tela 1.3 do fluxo de cadastro — confirmação dos dados informados."""

import tkinter as tk

import data.mock_data as mock_data
from config import CORES, FONTES, ESPACAMENTO
from components.widgets import botao_voltar, botao_primario, botao_secundario, indicador_progresso


class Passo3Frame(tk.Frame):
    """Terceiro passo do cadastro de lote: exibe um resumo dos dados
    preenchidos e simula a inserção do novo lote no banco de dados."""

    LINHAS_RESUMO = (
        ("Produtor", "produtor_nome", ""),
        ("Produto", "produto", ""),
        ("Quantidade", "quantidade", " kg"),
        ("Colheita", "data_colheita", ""),
        ("Validade", "validade", ""),
        ("Localização", "localizacao", ""),
    )

    def __init__(self, parent, app):
        super().__init__(parent, bg=CORES["fundo"])
        self.app = app

        topo = tk.Frame(self, bg=CORES["fundo"])
        botao_voltar(topo, lambda: self.app.show_frame("passo2")).pack(anchor="w")

        corpo = tk.Frame(self, bg=CORES["fundo"])

        indicador_progresso(corpo, 3).pack(anchor="w")
        tk.Label(
            corpo, text="Confirme os dados", bg=CORES["fundo"],
            fg=CORES["texto_principal"], font=FONTES["titulo"],
        ).pack(anchor="w", pady=(2, 20))

        self.card_resumo = tk.Frame(
            corpo, bg=CORES["branco"], highlightbackground=CORES["borda"],
            highlightthickness=1, padx=18, pady=4,
        )
        self.card_resumo.pack(fill="x")

        self.labels_valor = {}
        total_linhas = len(self.LINHAS_RESUMO)
        for indice, (rotulo, chave, _sufixo) in enumerate(self.LINHAS_RESUMO):
            linha = tk.Frame(self.card_resumo, bg=CORES["branco"])
            linha.pack(fill="x", pady=10)

            tk.Label(
                linha, text=rotulo, bg=CORES["branco"],
                fg=CORES["texto_secundario"], font=FONTES["corpo"],
            ).pack(side="left")

            label_valor = tk.Label(
                linha, text="", bg=CORES["branco"],
                fg=CORES["texto_principal"], font=(FONTES["corpo"][0], FONTES["corpo"][1], "bold"),
            )
            label_valor.pack(side="right")
            self.labels_valor[chave] = label_valor

            if indice < total_linhas - 1:
                tk.Frame(self.card_resumo, bg=CORES["borda"], height=1).pack(fill="x")

        rodape = tk.Frame(self, bg=CORES["fundo"])
        botao_primario(rodape, "Confirmar cadastro", self._confirmar).pack(fill="x", pady=(0, 8))
        botao_secundario(rodape, "Cancelar", self._cancelar).pack(fill="x")

        lateral = ESPACAMENTO["lateral"]
        topo.pack(fill="x", padx=lateral, pady=(ESPACAMENTO["topo"], 0))
        corpo.pack(fill="x", padx=lateral, pady=10)
        rodape.pack(fill="x", padx=lateral, pady=(0, 20), side="bottom")

    def on_show(self):
        """Lê os dados acumulados na sessão e atualiza o card de resumo."""
        for rotulo, chave, sufixo in self.LINHAS_RESUMO:
            valor = self.app.session.get(chave, "")
            self.labels_valor[chave].config(text=f"{valor}{sufixo}" if valor != "" else "")

    def _confirmar(self):
        """Simula o INSERT do novo lote na base de dados (dados mockados em memória).

        TODO (integração banco): substituir a simulação abaixo por:
            try:
                conn = get_connection()
                cur = conn.cursor()
                cur.execute('''
                    INSERT INTO lote_de_produto
                      (produto, cpf_cnpj_produtor, data_hora_cadastro,
                       quantidade, data_colheita, validade, localizacao_atual)
                    VALUES (%s, %s, NOW(), %s, %s, %s, %s)
                    RETURNING id_lote
                ''', (session["produto"], session["cpf_cnpj"], session["quantidade"],
                      session["data_colheita"], session["validade"], session["localizacao"]))
                session["id_gerado"] = cur.fetchone()[0]
                conn.commit()
            except Exception as e:
                conn.rollback()
                messagebox.showerror("Erro", "Não foi possível salvar o lote.")
                return
        """
        session = self.app.session
        novo_lote = {
            "id": mock_data.PROXIMO_ID,
            "produto": session.get("produto"),
            "produtor": session.get("produtor_nome"),
            "quantidade": session.get("quantidade"),
            "validade": session.get("validade"),
            "localizacao": session.get("localizacao"),
            # Recém-cadastrado: ainda não passou pela triagem, logo não tem classificação.
            "classificacao": "Não classificado",
        }
        mock_data.LOTES.append(novo_lote)
        session["id_gerado"] = mock_data.PROXIMO_ID
        mock_data.PROXIMO_ID += 1

        self.app.show_frame("sucesso")

    def _cancelar(self):
        """Descarta os dados preenchidos e volta ao menu principal."""
        self.app.limpar_session()
        self.app.show_frame("menu")

File: trabalho3/frontend/js/views/cadastro/passo3.js
// ── TELA 1.3: Cadastro — Passo 3 (Confirmação) ──

/**
 * Monta a estrutura HTML da tela de passo 3 e a insere em #app.
 */
function construirTelaPasso3() {
  const app = document.getElementById("app");

  const screen = document.createElement("div");
  screen.id = "screen-passo3";
  screen.className = "screen";

  screen.innerHTML = `
    <button class="btn-voltar" id="p3-voltar">← Voltar</button>
    <p class="progresso">Passo 3 de 3</p>
    <p class="titulo-tela">Confirme os dados</p>
    <p class="subtitulo-tela">Revise as informações antes de concluir o cadastro.</p>

    <div class="card-resumo">
      <div class="linha"><span class="chave">Produtor</span><span class="valor" id="p3-produtor"></span></div>
      <div class="linha"><span class="chave">Produto</span><span class="valor" id="p3-produto"></span></div>
      <div class="linha"><span class="chave">Quantidade</span><span class="valor" id="p3-quantidade"></span></div>
      <div class="linha"><span class="chave">Colheita</span><span class="valor" id="p3-colheita"></span></div>
      <div class="linha"><span class="chave">Validade</span><span class="valor" id="p3-validade"></span></div>
      <div class="linha"><span class="chave">Localização</span><span class="valor" id="p3-localizacao"></span></div>
    </div>

    <p class="label-erro" id="p3-erro-geral">Não foi possível salvar o lote. Tente novamente.</p>

    <button class="btn-primario" id="p3-confirmar">Confirmar cadastro</button>
    <button class="btn-secundario" id="p3-cancelar">Cancelar</button>
  `;

  app.appendChild(screen);

  screen.querySelector("#p3-voltar").addEventListener("click", () => {
    App.navegarPara("passo2");
  });

  screen.querySelector("#p3-cancelar").addEventListener("click", () => {
    App.limparSession();
    App.navegarPara("menu");
  });

  screen.querySelector("#p3-confirmar").addEventListener("click", confirmarCadastro);
}

/**
 * Exibe uma mensagem de erro geral na tela de confirmação.
 * @param {string} mensagem - Texto a ser exibido ao usuário.
 */
function exibirErroGeral(mensagem) {
  const erro = document.getElementById("p3-erro-geral");
  erro.textContent = mensagem;
  erro.classList.add("visivel");
}

/**
 * Envia o lote para a API (INSERT) e avança para a tela de sucesso.
 */
async function confirmarCadastro() {
  document.getElementById("p3-erro-geral").classList.remove("visivel");

  try {
    const res = await fetch("/api/lotes", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        produto:       App.session.produto,
        cpf_cnpj:      App.session.cpf_cnpj,
        quantidade:    App.session.quantidade,
        data_colheita: App.session.data_colheita,
        validade:      App.session.validade,
        localizacao:   App.session.localizacao,
      }),
    });

    if (!res.ok) {
      exibirErroGeral("Não foi possível salvar o lote. Tente novamente.");
      return;
    }

    const data = await res.json();
    App.session.id_gerado = data.id_lote;
    App.navegarPara("sucesso");
  } catch (erro) {
    exibirErroGeral("Não foi possível salvar o lote. Tente novamente.");
  }
}

construirTelaPasso3();

App.registrarTela("passo3", {
  /**
   * Preenche o card de resumo com os dados acumulados na sessão.
   */
  onShow() {
    const screen = document.getElementById("screen-passo3");
    screen.querySelector("#p3-produtor").textContent = App.session.produtor_nome ?? "";
    screen.querySelector("#p3-produto").textContent = App.session.produto ?? "";
    screen.querySelector("#p3-quantidade").textContent = `${App.session.quantidade ?? ""} kg`;
    screen.querySelector("#p3-colheita").textContent = App.session.data_colheita ?? "";
    screen.querySelector("#p3-validade").textContent = App.session.validade ?? "";
    screen.querySelector("#p3-localizacao").textContent = App.session.localizacao ?? "";
    screen.querySelector("#p3-erro-geral").classList.remove("visivel");
  },
});
